contracts/WETH10.sol

```diff
@@ -37,52 +37,64 @@ contract WETH10 {
                 keccak256(bytes("1")),
                 chainId,
                 address(this)));
+
+        // Trick to prevent transfers to the contract address without adding additional gas costs
+        balance[address(this)] = uint256(-1);
     }
 
     modifier lock() {
         require(unlocked == 1, "locked");
         unlocked = 0;
         _;
         unlocked = 1;
-    }    
+    }
+
+    function balanceOf(address account) external view returns (uint256) {
+        return account == address(this) ? 0 : balance[account];
+    }
 
     receive() external payable lock {
-        balanceOf[msg.sender] += msg.value;
+        require(balance[msg.sender] + msg.value >= msg.value, "overflow");
```

Comment:
Note that the `require` is only needed for transfers to the weth contract, as it is not possible to overflow any account by transferring ETH.

As such, it would be a bit cheaper to just `require(msg.sender != address(this), "!recipient")`.